### PR TITLE
 Node Centric Stat Hooks

### PR DESCRIPTION
**Pull request by xuanvietchu**
### Summary / Description

Added Node Centric Stat Hooks, example, and tests. Stats are computed with numpy for optimization. 
Produces:
        node_stats (Dict[int, Dict[str, float]]): Dictionary mapping node_id to statistics:
            - degree: Number of edges connected to the node in the current batch.
            - activity: Fraction of unique timesteps in which the node has appeared.
            - new_neighbors: Number of new neighbors in which the node encountered in the current batch.
            - lifetime: Time since the node was first seen.
            - time_since_last_seen: Time since the node was last seen.
            - appearances: Total number of unique timesteps the node has appeared in.
        node_macro_stats (Dict): Batch-level node statistics:
            - node_novelty: Fraction of tracked nodes in the batch that are appearing for the first time.
            - new_node_count: Number of tracked nodes in the batch that are appearing for the first time.
        edge_stats (Dict): Batch-level edge statistics:
            - edge_novelty: Fraction of new edges in the batch, that is not seen in previous batches.
            - edge_density: Edges this batch / possible edges based on unique nodes
            - new_edge_count: Number of new edges in the batch, that is not seen in previous batches.

**Related Issues:** #301 

#### Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking Change
- [x] Refactoring
- [ ] Documentation update

#### Test Evidence

Describe how this PR has been tested.

- [x] Unit tests
- [ ] Integration tests
- [ ] Performance tests

#### Questions / Discussion Points

List any areas where you’d like reviewer input or have open questions.

### CODE REVIEW

File: tgm/hooks/node_analytics.py
Comment (Jacob-Chmura) [inline tgm/hooks/node_analytics.py:1]: Can we update our documentation [here](https://github.com/tgm-team/tgm/blob/main/docs/api/hooks/hooks.md) to include `NodeAnalayticsHook` and `BatchAnalayticsHook`?

This will then automatically render the info here in our docs.

File: tgm/hooks/node_analytics.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+            - new_edge_count: Number of new edges in the batch, that is not seen in previous batches.
+    """
+
+    requires: Set[str] = set()
+    produces = {'node_stats', 'node_macro_stats', 'edge_stats'}
+
+    def __init__(
+        self,
+        tracked_nodes: Tensor,
+        num_nodes: int,
```
Comment (Jacob-Chmura) [inline tgm/hooks/node_analytics.py]: Nit:
```suggestion
        num_nodes: int
```

File: tgm/hooks/node_analytics.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+
+        # Concatenate src and dst to count all edge connections
+        edge_nodes = torch.cat([batch.src, batch.dst], dim=0)
+
+        # Vectorized degree computation
+        degrees = {}
+        nodes_np = nodes.cpu().numpy() if nodes.is_cuda else nodes.numpy()
+        edge_nodes_np = (
+            edge_nodes.cpu().numpy() if edge_nodes.is_cuda else edge_nodes.numpy()
+        )
```
Comment (Jacob-Chmura) [inline tgm/hooks/node_analytics.py]: Why is the conversion to numpy necessary? Can we just compute directly on the tensors?

File: tgm/hooks/node_analytics.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+        """Get all neighbors for each node in this batch."""
+        # Convert to numpy for faster iteration
+        nodes_np = nodes.cpu().numpy() if nodes.is_cuda else nodes.numpy()
+        nodes_set = set(int(n) for n in nodes_np)
+        neighbors: Dict[int, Set[int]] = {int(n): set() for n in nodes_set}
+
+        if batch.src is None or batch.dst is None:
+            return neighbors
+
+        # Convert to numpy arrays for faster iteration
```
Comment (Jacob-Chmura) [inline tgm/hooks/node_analytics.py:98]: I'm surprised that this would be faster than iterating the tensor? How much faster is it?

File: tgm/hooks/node_analytics.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+
+        if batch.node_ids is None or batch.node_ids.numel() == 0:
+            return node_stats
+
+        new_nodes = 0
+        for node in batch.node_ids:
+            node_id = int(node.item())
+            if node_id not in self._first_seen:
+                new_nodes += 1
+                self._seen_nodes.add(node_id)
```
Comment (Jacob-Chmura) [inline tgm/hooks/node_analytics.py]: Do we actually use `self._seen_nodes` anywhere? It appears that we can remove it altogether

File: tgm/hooks/node_analytics.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+            all_batch_nodes.append(batch.node_ids)
+
+        # No nodes in batch, return empty stats
+        if not all_batch_nodes:
+            node_batch_stats = self._compute_node_statistics(batch)
+            edge_batch_stats = self._compute_edge_statistics(batch)
+            batch.node_stats = {}  # type: ignore[attr-defined]
+            batch.node_macro_stats = node_batch_stats  # type: ignore[attr-defined]
+            batch.edge_stats = {**edge_batch_stats}  # type: ignore[attr-defined]
+            return batch
```
Comment (Jacob-Chmura) [inline tgm/hooks/node_analytics.py:235]: This is currently missing coverage, and it seems like a good edge case to consider. Can we add a test on an emty batch and ensure we get reasonable defaults out of it

File: tgm/hooks/node_analytics.py (hunk 2 above)
Comment (xuanvietchu) [reply tgm/hooks/node_analytics.py]: The reason behind most np cast is that I was profiling the code and realized that it was calling `".items()" a lot, which costs time. I just decided to cast everything to np and use np operations

I tested this by tracking 100 nodes on the example, and it's around 10% faster compared to the following:
`
degrees = {}

for node_val in nodes:
    degree = int((edge_nodes == node_val).sum().item())
    degrees[int(node_val)] = degree
`
 Also, for loops is usually faster on np arrays than tensors

File: tgm/hooks/node_analytics.py (hunk 3 above)
Comment (xuanvietchu) [reply tgm/hooks/node_analytics.py:98]: Each of these .numpy() casting sections contributed to around 10% of the speed (100 nodes tracked)

File: tgm/hooks/node_analytics.py (hunk 4 above)
Comment (xuanvietchu) [reply tgm/hooks/node_analytics.py]: This is actually an oversight on my end. That's part of my outdated code. Thanks!

File: tgm/hooks/node_analytics.py (hunk 2 above)
Comment (Jacob-Chmura) [reply tgm/hooks/node_analytics.py]: What if we vectorize the degree counting, something like:

```python
unique_nodes, inv = torch.unique(edge_nodes, return_inverse=True)
counts = torch.bincount(inv)
deg_map = dict(zip(unique_nodes.tolist(), counts.tolist()))
return {int(n): deg_map.get(int(n), 0) for n in nodes.tolist()}
```

File: tgm/hooks/node_analytics.py [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+
+class NodeAnalyticsHook(StatefulHook):
+    """Compute node-centric statistics for a specific set of tracked nodes.
+
+    This hook maintains state across batches to compute temporal statistics
+    for a specified set of nodes.
+
+    Args:
+        tracked_nodes (Tensor): 1D tensor of node IDs to track statistics for.
+        compute_moving_avg (bool): If True, compute exponential moving averages.
```
Comment (ntgbaoo) [inline tgm/hooks/node_analytics.py]: Builds doc test failed due to `compute_moving_avg` documented but not used. Do we need this attribute?

File: tgm/hooks/node_analytics.py [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+            - lifetime: Time since the node was first seen.
+            - time_since_last_seen: Time since the node was last seen.
+            - appearances: Total number of unique timesteps the node has appeared in.
+        node_macro_stats (Dict): Batch-level node statistics:
+            - node_novelty: Fraction of tracked nodes in the batch that are appearing for the first time.
+            - new_node_count: Number of tracked nodes in the batch that are appearing for the first time.
+        edge_stats (Dict): Batch-level edge statistics:
+            - edge_novelty: Fraction of new edges in the batch, that is not seen in previous batches.
+            - edge_density: Edges this batch / possible edges based on unique nodes
+            - new_edge_count: Number of new edges in the batch, that is not seen in previous batches.
```
Comment (ntgbaoo) [inline tgm/hooks/node_analytics.py:38]: nit: The name of the hook, `NodeAnalyticsHook`, is currently a bit misleading to users. Why does node analytics return `edge_stats`. Maybe `NodeLevelAnalyticsHook` or `NodeCentricAnalyticsHook`

File: tgm/hooks/node_analytics.py [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+    """
+
+    requires: Set[str] = set()
+    produces = {'node_stats', 'node_macro_stats', 'edge_stats'}
+
+    def __init__(self, tracked_nodes: Tensor, num_nodes: int) -> None:
```
Comment (ntgbaoo) [inline tgm/hooks/node_analytics.py:44]: we are missing documentation for `num_nodes` in the comment. That is also why the build doc check failed.

File: tgm/hooks/node_analytics.py (hunk 2 above)
Comment (ntgbaoo) [reply tgm/hooks/node_analytics.py]: In any case, I think it is ideal to keep everything on the cuda and avoid overhead from converting from `cpu` to `cuda`, and vice versa. I think you can try out Jacob's suggestion, another way to do it for your reference:
```python
index =  torch.cat([src,dst])
degree = torch.zeros(number_of_node)
degree = degree.scatter_add_(0, index, degree.new_ones((index.size(0))))
```

I just wrote it up from the terminal. You may need to adapt to make it work in your case.

File: tgm/hooks/node_analytics.py (hunk 3 above)
Comment (benjaminnNgo) [reply tgm/hooks/node_analytics.py]: If we want to vectorize this, we may need to convert `edge_index` into adj matrix and aggregate to get a list of neighbours for each node at the end?

Comment (ntgbaoo) [reply tgm/hooks/node_analytics.py:98]: But I think it makes sense why this is faster. By doing it this way: converting from cuda to cpu happens once, while in the case of looping over the tensor, the conversion happens for every iteration (since int(x..item()) will transfer from cuda to cpu by default). 

If we really want to work on tensor only, we may need to convert `edge_index` into adj matrix and aggregate to get a list of neighbours for each node at the end?

**Review by benjaminnNgo — COMMENTED** [review]
Hey Viet!

Thank you for the PR. In general, it looks good to me. One major concern is just the doc build failed (see my suggestion). I left a minor suggestion on vectorizing the degree operation.

File: tgm/hooks/node_analytics.py (hunk 2 above)
Comment (xuanvietchu) [reply tgm/hooks/node_analytics.py]: > ```python
> unique_nodes, inv = torch.unique(edge_nodes, return_inverse=True)
> ```

I added this to the code base, turns out ```.unique()``` and ```.bincount()``` is a bit faster (~7%) than the numpy casting and ```(edge_nodes_np == node_val).sum()``` despite of the ```toList()``` conversions

File: tgm/hooks/node_analytics.py (hunk 6 above)
Comment (xuanvietchu) [reply tgm/hooks/node_analytics.py]: These are is my leftover code I missed during clean up, thanks!

File: tgm/hooks/node_analytics.py (hunk 8 above)
Comment (xuanvietchu) [reply tgm/hooks/node_analytics.py:44]: I've updated the doc; you can try it again.

File: examples/analytics/node_analytics_example.py [hunk 9]
```diff
@@ -21,7 +21,7 @@
 parser.add_argument('--dataset', type=str, default='tgbn-trade', help='Dataset name')
 parser.add_argument('--bsize', type=int, default=200, help='batch size')
 parser.add_argument(
-    '--num-tracked', type=int, default=1, help='Number of nodes to track'
+    '--num-tracked', type=int, default=500, help='Number of nodes to track'
```
Comment (Jacob-Chmura) [inline examples/analytics/node_analytics_example.py]: I haven't ran with 500 nodes tracked, but for the purposes of an example, I think something smaller (e.g. 10) might be sufficient

File: tgm/hooks/node_analytics.py [hunk 10]
```diff
@@ -11,15 +11,15 @@
 logger = _get_logger(__name__)
 
 
-class NodeAnalyticsHook(StatefulHook):
+class NodeCentricAnalyticsHook(StatefulHook):
```
Comment (Jacob-Chmura) [inline tgm/hooks/node_analytics.py]: I actually prefer `NodeAnalyticsHook` as it's shorter and conveys pretty much the same thing as `NodeCentricAnalyticsHook`
